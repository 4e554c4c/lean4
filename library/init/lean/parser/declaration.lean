/-
Copyright (c) 2018 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Sebastian Ullrich

Parsers for commands that declare things
-/

prelude
import init.lean.parser.term

namespace lean
namespace parser

open combinators monadParsec
open parser.hasTokens parser.hasView

instance termParserCommandParserCoe : hasCoe termParser commandParser :=
⟨λ p, adaptReader coe $ p.run⟩

namespace «command»

local postfix `?`:10000 := optional
local postfix *:10000 := combinators.many
local postfix +:10000 := combinators.many1

@[derive hasTokens hasView]
def docComment.parser : commandParser :=
node! docComment ["/--", doc: raw $ many' (notFollowedBy (str "-/") *> any), "-/"]

@[derive hasTokens hasView]
def attrInstance.parser : commandParser :=
-- use `rawIdent` because of attribute names such as `instance`
node! attrInstance [name: rawIdent.parser, args: (term.parser maxPrec)*]

@[derive hasTokens hasView]
def declAttributes.parser : commandParser :=
-- TODO(Sebastian): custom attribute parsers
node! declAttributes ["@[", attrs: sepBy1 attrInstance.parser (symbol ","), "]"]

setOption class.instanceMaxDepth 300
@[derive hasTokens hasView]
def declModifiers.parser : commandParser :=
node! declModifiers [
  docComment: docComment.parser?,
  attrs: declAttributes.parser?,
  visibility: nodeChoice! visibility {"private", "protected"}?,
  «noncomputable»: (symbol "noncomputable")?,
  «unsafe»: (symbol "unsafe")?,
]

@[derive hasTokens hasView]
def declSig.parser : commandParser :=
node! declSig [
  params: term.bracketedBinders.parser,
  type: term.typeSpec.parser,
]

@[derive hasTokens hasView]
def optDeclSig.parser : commandParser :=
node! optDeclSig [
  params: term.bracketedBinders.parser,
  type: term.optType.parser,
]

@[derive hasTokens hasView]
def equation.parser : commandParser :=
node! equation ["|", lhs: (term.parser maxPrec)+, ":=", rhs: term.parser]

@[derive hasTokens hasView]
def declVal.parser : commandParser :=
nodeChoice! declVal {
  simple: node! simpleDeclVal [":=", body: term.parser],
  emptyMatch: symbol ".",
  «match»: equation.parser+
}

@[derive hasTokens hasView]
def inferModifier.parser : commandParser :=
nodeChoice! inferModifier {
  relaxed: try $ node! relaxedInferModifier ["{", "}"],
  strict: try $ node! strictInferModifier ["(", ")"],
}

@[derive hasTokens hasView]
def introRule.parser : commandParser :=
node! introRule [
  "|",
  name: ident.parser,
  inferMod: inferModifier.parser?,
  sig: optDeclSig.parser,
]

@[derive hasTokens hasView]
def structBinderContent.parser : commandParser :=
node! structBinderContent [
  ids: ident.parser+,
  inferMod: inferModifier.parser?,
  sig: optDeclSig.parser,
  default: term.binderDefault.parser?,
]

@[derive hasTokens hasView]
def structureFieldBlock.parser : commandParser :=
nodeChoice! structureFieldBlock {
  explicit: node! structExplicitBinder ["(", content: nodeChoice! structExplicitBinderContent {
    «notation»: command.notationLike.parser,
    other: structBinderContent.parser
  }, right: symbol ")"],
  implicit: node! structImplicitBinder ["{", content: structBinderContent.parser, "}"],
  strictImplicit: node! strictImplicitBinder ["⦃", content: structBinderContent.parser, "⦄"],
  instImplicit: node! instImplicitBinder ["[", content: structBinderContent.parser, "]"],
}

@[derive hasTokens hasView]
def oldUnivParams.parser : commandParser :=
node! oldUnivParams ["{", ids: ident.parser+, "}"]

@[derive parser.hasTokens parser.hasView]
def identUnivParams.parser : commandParser :=
node! identUnivParams [
  id: ident.parser,
  univParams: node! univParams [".{", params: ident.parser+, "}"]?
]

@[derive hasTokens hasView]
def structure.parser : commandParser :=
node! «structure» [
  keyword: nodeChoice! structureKw {"structure", "class"},
  oldUnivParams: oldUnivParams.parser?,
  name: identUnivParams.parser,
  sig: optDeclSig.parser,
  «extends»: node! «extends» ["extends", parents: sepBy1 term.parser (symbol ",")]?,
  ":=",
  ctor: node! structureCtor [name: ident.parser, inferMod: inferModifier.parser?, "::"]?,
  fieldBlocks: structureFieldBlock.parser*,
]

@[derive hasTokens hasView]
def declaration.parser : commandParser :=
node! declaration [
  modifiers: declModifiers.parser,
  inner: nodeChoice! declaration.inner {
    «defLike»: node! «defLike» [
      kind: nodeChoice! defLike.kind {"def", "abbreviation", "abbrev", "theorem", "constant"},
      oldUnivParams: oldUnivParams.parser?,
      name: identUnivParams.parser, sig: optDeclSig.parser, val: declVal.parser],
    «instance»: node! «instance» ["instance", name: identUnivParams.parser?, sig: declSig.parser, val: declVal.parser],
    «example»: node! «example» ["example", sig: declSig.parser, val: declVal.parser],
    «axiom»: node! «axiom» [
      kw: nodeChoice! constantKeyword {"axiom"},
      name: identUnivParams.parser,
      sig: declSig.parser],
    «inductive»: node! «inductive» [try [«class»: (symbol "class")?, "inductive"],
      oldUnivParams: oldUnivParams.parser?,
      name: identUnivParams.parser,
      sig: optDeclSig.parser,
      localNotation: notationLike.parser?,
      introRules: introRule.parser*],
    «structure»: structure.parser,
  }
]

end «command»
end parser
end lean
