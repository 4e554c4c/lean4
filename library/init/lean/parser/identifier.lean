/-
Copyright (c) 2018 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
prelude
import init.data.char.basic init.lean.parser.parsec

namespace lean

def is_greek (c : char) : bool :=
0x391 ≤ c.val && c.val ≤ 0x3dd

def is_letter_like (c : char) : bool :=
(0x3b1  ≤ c.val && c.val ≤ 0x3c9 && c.val ≠ 0x3bb) ||                  -- Lower greek, but lambda
(0x391  ≤ c.val && c.val ≤ 0x3A9 && c.val ≠ 0x3A0 && c.val ≠ 0x3A3) || -- Upper greek, but Pi and Sigma
(0x3ca  ≤ c.val && c.val ≤ 0x3fb) ||                                   -- Coptic letters
(0x1f00 ≤ c.val && c.val ≤ 0x1ffe) ||                                  -- Polytonic Greek Extended Character Set
(0x2100 ≤ c.val && c.val ≤ 0x214f) ||                                  -- Letter like block
(0x1d49c ≤ c.val && c.val ≤ 0x1d59f)                                   -- Latin letters, Script, Double-struck, Fractur

def is_sub_script_alnum (c : char) : bool :=
(0x207f ≤ c.val && c.val ≤ 0x2089) || -- n superscript and numberic subscripts
(0x2090 ≤ c.val && c.val ≤ 0x209c) ||
(0x1d62 ≤ c.val && c.val ≤ 0x1d6a)

def is_id_first (c : char) : bool :=
c.is_alpha || c = '_' || is_letter_like c

def is_id_rest (c : char) : bool :=
c.is_alphanum || c = '_' || c = '\'' || is_letter_like c || is_sub_script_alnum c

def id_begin_escape := '«'
def id_end_escape   := '»'
def is_id_begin_escape (c : char) : bool :=
c = id_begin_escape
def is_id_end_escape (c : char) : bool :=
c = id_end_escape

namespace parser
open monad_parsec

def id_part_default : parsec' string :=
do c ← satisfy is_id_first,
   take_while_cont is_id_rest (to_string c)

def id_part_escaped : parsec' string :=
ch id_begin_escape *> take_until1 is_id_end_escape <* ch id_end_escape

def id_part : parsec' string :=
cond is_id_begin_escape
  id_part_escaped
  id_part_default

def identifier : parsec' name :=
(try $ do s  ← id_part,
       foldl name.mk_string (mk_simple_name s) (ch '.' *> id_part)) <?> "identifier"

def c_identifier : parsec' string :=
(try $ do c ← satisfy (λ c, c.is_alpha || c = '_'),
       take_while_cont (λ c, c.is_alphanum || c = '_') (to_string c)) <?> "C identifier"

def cpp_identifier : parsec' string :=
(try $ do n ← c_identifier,
       ns ← many ((++) <$> str "::" <*> c_identifier),
       pure $ string.join (n::ns)) <?> "C++ identifier"

end parser
end lean
