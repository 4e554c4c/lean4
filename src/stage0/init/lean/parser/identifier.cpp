// Lean compiler output
// Module: init.lean.parser.identifier
// Imports: init.data.char.basic
#include "runtime/object.h"
#include "runtime/apply.h"
typedef lean::object obj;    typedef lean::usize  usize;
typedef lean::uint8  uint8;  typedef lean::uint16 uint16;
typedef lean::uint32 uint32; typedef lean::uint64 uint64;
#if defined(__clang__)
#pragma clang diagnostic ignored "-Wunused-parameter"
#pragma clang diagnostic ignored "-Wunused-label"
#elif defined(__GNUC__) && !defined(__CLANG__)
#pragma GCC diagnostic ignored "-Wunused-parameter"
#pragma GCC diagnostic ignored "-Wunused-label"
#pragma GCC diagnostic ignored "-Wunused-but-set-variable"
#endif
extern "C" {
uint8 l_Lean_isSubScriptAlnum(uint32);
obj* l_Lean_isIdRest___boxed(obj*);
uint8 l_Char_isAlphanum(uint32);
obj* l_Lean_isLetterLike___boxed(obj*);
uint8 l_Lean_isIdRest(uint32);
uint8 l_Lean_isIdEndEscape(uint32);
obj* l_Lean_isIdBeginEscape___boxed(obj*);
uint8 l_Char_isAlpha(uint32);
obj* l_Lean_isIdFirst___boxed(obj*);
uint8 l_Lean_isLetterLike(uint32);
uint8 l_Lean_isIdBeginEscape(uint32);
uint8 l_UInt32_decLe(uint32, uint32);
uint32 l_Lean_idBeginEscape;
uint8 l_UInt32_decEq(uint32, uint32);
obj* l_Lean_isGreek___boxed(obj*);
obj* l_Lean_isIdEndEscape___boxed(obj*);
uint32 l_Lean_idEndEscape;
uint8 l_Lean_isIdFirst(uint32);
uint8 l_Lean_isGreek(uint32);
obj* l_Lean_isSubScriptAlnum___boxed(obj*);
uint8 l_Lean_isGreek(uint32 x_1) {
_start:
{
uint32 x_2; uint8 x_3; 
x_2 = 913;
x_3 = x_2 <= x_1;
if (x_3 == 0)
{
uint8 x_4; 
x_4 = 0;
return x_4;
}
else
{
uint32 x_5; uint8 x_6; 
x_5 = 989;
x_6 = x_1 <= x_5;
return x_6;
}
}
}
obj* l_Lean_isGreek___boxed(obj* x_1) {
_start:
{
uint32 x_2; uint8 x_3; obj* x_4; 
x_2 = lean::unbox_uint32(x_1);
lean::dec(x_1);
x_3 = l_Lean_isGreek(x_2);
x_4 = lean::box(x_3);
return x_4;
}
}
uint8 l_Lean_isLetterLike(uint32 x_1) {
_start:
{
uint32 x_2; uint8 x_3; uint8 x_19; uint8 x_27; uint8 x_35; uint8 x_59; 
x_2 = 945;
x_59 = x_2 <= x_1;
if (x_59 == 0)
{
uint8 x_60; 
x_60 = 0;
x_35 = x_60;
goto block_58;
}
else
{
uint32 x_61; uint8 x_62; 
x_61 = 969;
x_62 = x_1 <= x_61;
if (x_62 == 0)
{
uint8 x_63; 
x_63 = 0;
x_35 = x_63;
goto block_58;
}
else
{
uint32 x_64; uint8 x_65; 
x_64 = 955;
x_65 = x_1 == x_64;
if (x_65 == 0)
{
uint8 x_66; 
x_66 = 1;
return x_66;
}
else
{
uint8 x_67; 
x_67 = 0;
x_35 = x_67;
goto block_58;
}
}
}
block_18:
{
uint32 x_4; uint8 x_5; 
x_4 = 8448;
x_5 = x_4 <= x_1;
if (x_5 == 0)
{
if (x_3 == 0)
{
uint32 x_6; uint8 x_7; 
x_6 = 119964;
x_7 = x_6 <= x_1;
if (x_7 == 0)
{
return x_3;
}
else
{
uint32 x_8; uint8 x_9; 
x_8 = 120223;
x_9 = x_1 <= x_8;
return x_9;
}
}
else
{
uint8 x_10; 
x_10 = 1;
return x_10;
}
}
else
{
uint32 x_11; uint8 x_12; 
x_11 = 8527;
x_12 = x_1 <= x_11;
if (x_12 == 0)
{
uint32 x_13; uint8 x_14; 
x_13 = 119964;
x_14 = x_13 <= x_1;
if (x_14 == 0)
{
return x_12;
}
else
{
uint32 x_15; uint8 x_16; 
x_15 = 120223;
x_16 = x_1 <= x_15;
return x_16;
}
}
else
{
uint8 x_17; 
x_17 = 1;
return x_17;
}
}
}
block_26:
{
uint32 x_20; uint8 x_21; 
x_20 = 7936;
x_21 = x_20 <= x_1;
if (x_21 == 0)
{
if (x_19 == 0)
{
x_3 = x_19;
goto block_18;
}
else
{
uint8 x_22; 
x_22 = 1;
return x_22;
}
}
else
{
uint32 x_23; uint8 x_24; 
x_23 = 8190;
x_24 = x_1 <= x_23;
if (x_24 == 0)
{
x_3 = x_24;
goto block_18;
}
else
{
uint8 x_25; 
x_25 = 1;
return x_25;
}
}
}
block_34:
{
uint32 x_28; uint8 x_29; 
x_28 = 970;
x_29 = x_28 <= x_1;
if (x_29 == 0)
{
if (x_27 == 0)
{
x_19 = x_27;
goto block_26;
}
else
{
uint8 x_30; 
x_30 = 1;
return x_30;
}
}
else
{
uint32 x_31; uint8 x_32; 
x_31 = 1019;
x_32 = x_1 <= x_31;
if (x_32 == 0)
{
x_19 = x_32;
goto block_26;
}
else
{
uint8 x_33; 
x_33 = 1;
return x_33;
}
}
}
block_58:
{
uint32 x_36; uint8 x_37; 
x_36 = 913;
x_37 = x_36 <= x_1;
if (x_37 == 0)
{
if (x_35 == 0)
{
x_27 = x_35;
goto block_34;
}
else
{
uint32 x_38; uint8 x_39; 
x_38 = 928;
x_39 = x_1 == x_38;
if (x_39 == 0)
{
uint32 x_40; uint8 x_41; 
x_40 = 931;
x_41 = x_1 == x_40;
if (x_41 == 0)
{
uint8 x_42; 
x_42 = 1;
return x_42;
}
else
{
uint8 x_43; 
x_43 = 0;
x_27 = x_43;
goto block_34;
}
}
else
{
uint8 x_44; 
x_44 = 1;
return x_44;
}
}
}
else
{
uint32 x_45; uint8 x_46; 
x_45 = 937;
x_46 = x_1 <= x_45;
if (x_46 == 0)
{
if (x_35 == 0)
{
x_27 = x_35;
goto block_34;
}
else
{
uint32 x_47; uint8 x_48; 
x_47 = 931;
x_48 = x_1 == x_47;
if (x_48 == 0)
{
uint8 x_49; 
x_49 = 1;
return x_49;
}
else
{
uint8 x_50; 
x_50 = 0;
x_27 = x_50;
goto block_34;
}
}
}
else
{
uint32 x_51; uint8 x_52; 
x_51 = 928;
x_52 = x_1 == x_51;
if (x_52 == 0)
{
uint32 x_53; uint8 x_54; 
x_53 = 931;
x_54 = x_1 == x_53;
if (x_54 == 0)
{
uint8 x_55; 
x_55 = 1;
return x_55;
}
else
{
uint8 x_56; 
x_56 = 0;
x_27 = x_56;
goto block_34;
}
}
else
{
if (x_35 == 0)
{
x_27 = x_35;
goto block_34;
}
else
{
uint8 x_57; 
x_57 = 1;
return x_57;
}
}
}
}
}
}
}
obj* l_Lean_isLetterLike___boxed(obj* x_1) {
_start:
{
uint32 x_2; uint8 x_3; obj* x_4; 
x_2 = lean::unbox_uint32(x_1);
lean::dec(x_1);
x_3 = l_Lean_isLetterLike(x_2);
x_4 = lean::box(x_3);
return x_4;
}
}
uint8 l_Lean_isSubScriptAlnum(uint32 x_1) {
_start:
{
uint32 x_2; uint8 x_3; uint8 x_19; 
x_2 = 8319;
x_19 = x_2 <= x_1;
if (x_19 == 0)
{
uint8 x_20; 
x_20 = 0;
x_3 = x_20;
goto block_18;
}
else
{
uint32 x_21; uint8 x_22; 
x_21 = 8329;
x_22 = x_1 <= x_21;
if (x_22 == 0)
{
x_3 = x_22;
goto block_18;
}
else
{
uint8 x_23; 
x_23 = 1;
return x_23;
}
}
block_18:
{
uint32 x_4; uint8 x_5; 
x_4 = 8336;
x_5 = x_4 <= x_1;
if (x_5 == 0)
{
if (x_3 == 0)
{
uint32 x_6; uint8 x_7; 
x_6 = 7522;
x_7 = x_6 <= x_1;
if (x_7 == 0)
{
return x_3;
}
else
{
uint32 x_8; uint8 x_9; 
x_8 = 7530;
x_9 = x_1 <= x_8;
return x_9;
}
}
else
{
uint8 x_10; 
x_10 = 1;
return x_10;
}
}
else
{
uint32 x_11; uint8 x_12; 
x_11 = 8348;
x_12 = x_1 <= x_11;
if (x_12 == 0)
{
uint32 x_13; uint8 x_14; 
x_13 = 7522;
x_14 = x_13 <= x_1;
if (x_14 == 0)
{
return x_12;
}
else
{
uint32 x_15; uint8 x_16; 
x_15 = 7530;
x_16 = x_1 <= x_15;
return x_16;
}
}
else
{
uint8 x_17; 
x_17 = 1;
return x_17;
}
}
}
}
}
obj* l_Lean_isSubScriptAlnum___boxed(obj* x_1) {
_start:
{
uint32 x_2; uint8 x_3; obj* x_4; 
x_2 = lean::unbox_uint32(x_1);
lean::dec(x_1);
x_3 = l_Lean_isSubScriptAlnum(x_2);
x_4 = lean::box(x_3);
return x_4;
}
}
uint8 l_Lean_isIdFirst(uint32 x_1) {
_start:
{
uint8 x_2; 
x_2 = l_Char_isAlpha(x_1);
if (x_2 == 0)
{
uint32 x_3; uint8 x_4; 
x_3 = 95;
x_4 = x_1 == x_3;
if (x_4 == 0)
{
uint8 x_5; 
x_5 = l_Lean_isLetterLike(x_1);
return x_5;
}
else
{
uint8 x_6; 
x_6 = 1;
return x_6;
}
}
else
{
uint8 x_7; 
x_7 = 1;
return x_7;
}
}
}
obj* l_Lean_isIdFirst___boxed(obj* x_1) {
_start:
{
uint32 x_2; uint8 x_3; obj* x_4; 
x_2 = lean::unbox_uint32(x_1);
lean::dec(x_1);
x_3 = l_Lean_isIdFirst(x_2);
x_4 = lean::box(x_3);
return x_4;
}
}
uint8 l_Lean_isIdRest(uint32 x_1) {
_start:
{
uint8 x_2; 
x_2 = l_Char_isAlphanum(x_1);
if (x_2 == 0)
{
uint32 x_3; uint8 x_4; 
x_3 = 95;
x_4 = x_1 == x_3;
if (x_4 == 0)
{
uint32 x_5; uint8 x_6; 
x_5 = 39;
x_6 = x_1 == x_5;
if (x_6 == 0)
{
uint8 x_7; 
x_7 = l_Lean_isLetterLike(x_1);
if (x_7 == 0)
{
uint8 x_8; 
x_8 = l_Lean_isSubScriptAlnum(x_1);
return x_8;
}
else
{
uint8 x_9; 
x_9 = 1;
return x_9;
}
}
else
{
uint8 x_10; 
x_10 = 1;
return x_10;
}
}
else
{
uint8 x_11; 
x_11 = 1;
return x_11;
}
}
else
{
uint8 x_12; 
x_12 = 1;
return x_12;
}
}
}
obj* l_Lean_isIdRest___boxed(obj* x_1) {
_start:
{
uint32 x_2; uint8 x_3; obj* x_4; 
x_2 = lean::unbox_uint32(x_1);
lean::dec(x_1);
x_3 = l_Lean_isIdRest(x_2);
x_4 = lean::box(x_3);
return x_4;
}
}
uint32 _init_l_Lean_idBeginEscape() {
_start:
{
uint32 x_1; 
x_1 = 171;
return x_1;
}
}
uint32 _init_l_Lean_idEndEscape() {
_start:
{
uint32 x_1; 
x_1 = 187;
return x_1;
}
}
uint8 l_Lean_isIdBeginEscape(uint32 x_1) {
_start:
{
uint32 x_2; uint8 x_3; 
x_2 = l_Lean_idBeginEscape;
x_3 = x_1 == x_2;
if (x_3 == 0)
{
uint8 x_4; 
x_4 = 0;
return x_4;
}
else
{
uint8 x_5; 
x_5 = 1;
return x_5;
}
}
}
obj* l_Lean_isIdBeginEscape___boxed(obj* x_1) {
_start:
{
uint32 x_2; uint8 x_3; obj* x_4; 
x_2 = lean::unbox_uint32(x_1);
lean::dec(x_1);
x_3 = l_Lean_isIdBeginEscape(x_2);
x_4 = lean::box(x_3);
return x_4;
}
}
uint8 l_Lean_isIdEndEscape(uint32 x_1) {
_start:
{
uint32 x_2; uint8 x_3; 
x_2 = l_Lean_idEndEscape;
x_3 = x_1 == x_2;
if (x_3 == 0)
{
uint8 x_4; 
x_4 = 0;
return x_4;
}
else
{
uint8 x_5; 
x_5 = 1;
return x_5;
}
}
}
obj* l_Lean_isIdEndEscape___boxed(obj* x_1) {
_start:
{
uint32 x_2; uint8 x_3; obj* x_4; 
x_2 = lean::unbox_uint32(x_1);
lean::dec(x_1);
x_3 = l_Lean_isIdEndEscape(x_2);
x_4 = lean::box(x_3);
return x_4;
}
}
obj* initialize_init_data_char_basic(obj*);
static bool _G_initialized = false;
obj* initialize_init_lean_parser_identifier(obj* w) {
if (_G_initialized) return w;
_G_initialized = true;
if (lean::io_result_is_error(w)) return w;
w = initialize_init_data_char_basic(w);
if (lean::io_result_is_error(w)) return w;
l_Lean_idBeginEscape = _init_l_Lean_idBeginEscape();
l_Lean_idEndEscape = _init_l_Lean_idEndEscape();
return w;
}
}
