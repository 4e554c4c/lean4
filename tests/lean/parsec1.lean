import init.lean.parser.identifier
open Lean.Parser
open Lean.Parser.MonadParsec

def test {α} [DecidableEq α] (p : Parsec' α) (s : String) (e : α) : IO Unit :=
match Parsec.parse p s with
| Except.ok a    := if a = e then pure () else IO.println "unexpected Result"
| Except.error e := IO.println e

def testFailure {α} (p : Parsec' α) (s : String) : IO Unit :=
match Parsec.parse p s with
| Except.ok a    := IO.println "unexpected success"
| Except.error e := pure ()

def showResult {α} [HasToString α] (p : Parsec' α) (s : String) : IO Unit :=
match Parsec.parse p s with
| Except.ok a    := IO.println "Result: " *> IO.println (repr $ toString a)
| Except.error e := IO.println e

#exit -- Disabled until we implement new VM

#eval test (ch 'a') "a" 'a'
#eval test any "a" 'a'
#eval test any "b" 'b'
#eval test (str "foo" <|> str "bla" <|> str "boo") "bla" "bla"
#eval test ((str "foo" *> str "foo") <|> str "bla" <|> str "boo") "bla" "bla"
#eval testFailure ((str "foo" *> str "foo") <|> str "foo2" <|> str "boo") "foo2"
#eval test (try (str "foo" *> str "foo") <|> str "foo2" <|> str "boo") "foo2" "foo2"
#eval test num "1000" 1000
#eval test (do n ← num, whitespace, m ← num, pure (n, m)) "1000 200" (1000, 200)
#eval test (do n ← num, whitespace, m ← num, pure (n, m)) "1000      200" (1000, 200)
#eval test (do n ← lexeme num, m ← num, pure (n, m)) "1000 200" (1000, 200)
#eval test (whitespace *> Prod.mk <$> (lexeme num) <*> (lexeme num)) "    1000 200    " (1000, 200)
#eval test (whitespace *> Prod.mk <$> (lexeme num) <*> (lexeme num) <* eoi) "    1000 200    " (1000, 200)
#eval testFailure (whitespace *> Prod.mk <$> (lexeme num) <*> num <* eoi) "    1000 200    "

#eval testFailure ((ch 'a' *> ch 'b') <|> (ch 'a' *> ch 'c')) "ac"
#eval test ((lookahead (str "ab") *> ch 'a' *> ch 'b') <|> (ch 'a' *> ch 'c')) "ac" 'c'
#eval test (str "ab" *> pure () <|> (ch 'a' *> ch 'c' *> pure ())) "ac" ()
#eval test (try (ch 'a' *> ch 'b') <|> (ch 'a' *> ch 'c')) "ac" 'c'
#eval test (lookahead (ch 'a')) "abc" 'a'
#eval testFailure (notFollowedBy (lookahead (ch 'a'))) "abc"

def symbol (c : Char) : Parsec' Char :=
lexeme (ch c) <?> repr c

def paren {α} (p : Parsec' α) : Parsec' α :=
symbol '(' *> lexeme p <* symbol ')'

#eval test (paren num) "(   10 )" 10
#eval test (paren num) "(12)" 12
#eval test (paren num) "(0)" 0
#eval test (paren num) "(0    )" 0

def var : Parsec' String :=
do c ← satisfy (λ a, a.isAlpha || a = '_'),
   r ← lexeme $ takeWhile (λ a, a.isDigit || a.isAlpha || a = '_'),
   pure (c.toString ++ r)

#eval test var "abc" "abc"
#eval test var "_a1Bc" "_a1Bc"
#eval test (paren var) "(_a1Bc )" "_a1Bc"
#eval testFailure var "1A1Bc"
#eval testFailure var "*_a1Bc"
#eval test var "abc$" "abc"

open Lean

#eval test identifier "«!!aaa».b1'" (mkStrName (mkSimpleName "!!aaa") "b1'")
#eval test identifier "a" (mkSimpleName "a")
#eval test identifier "a'" (mkSimpleName "a'")
#eval test identifier "_" (mkSimpleName "_")
#eval test identifier "_a1" (mkSimpleName "_a1")
#eval test identifier "aaa.bbb._αc" (mkStrName (mkStrName (mkSimpleName "aaa") "bbb") "_αc")
#eval test identifier "«!a!aa».b12.ccc" (mkStrName (mkStrName (mkSimpleName "!a!aa") "b12") "ccc")
#eval testFailure identifier "1A1Bc"
#eval testFailure identifier "!"
#eval testFailure identifier "1"
#eval testFailure identifier "'a"
#eval testFailure identifier ""
#eval testFailure identifier " "

#eval test parseStringLiteral "\"abc\"" "abc"
#eval test parseStringLiteral "\"\\\\abc\"" "\\abc"
#eval test parseStringLiteral "\"\"" ""
#eval test parseStringLiteral "\"\\\"\"" "\""
#eval test parseStringLiteral "\"\\\'\"" "\'"
#eval test parseStringLiteral "\"\\n\"" "\n"
#eval test parseStringLiteral "\"\\t\"" "\t"
#eval test parseStringLiteral "\"\\x4e\"" "N"
#eval test parseStringLiteral "\"\\x4E\"" "N"
#eval test parseStringLiteral "\"\\x7D\"" "}"
#eval test parseStringLiteral "\"\\u03b1\\u03b1\"" "αα"
#eval testFailure parseStringLiteral "\"abc"
#eval testFailure parseStringLiteral "\"\\abc\""
#eval testFailure parseStringLiteral "\"\\x4z\""
#eval testFailure parseStringLiteral "\"\\x4\""
#eval testFailure parseStringLiteral "\"\\u03b\\u03b1\""
#eval testFailure parseStringLiteral "\"\\u03bz\\u03b1\""

def parseInstrPp : Parsec' String :=
do cmd ← Lean.Ir.parseInstr,
   pure $ toString cmd

#eval test parseInstrPp "x :    UInt32 :=  10" "x : UInt32 := 10"
#eval test parseInstrPp "x : Bool:=not y" "x : Bool := not y"
#eval test parseInstrPp "x : Bool := and z   y" "x : Bool := and z y"
#eval test parseInstrPp "x := call f z w" "x := call f z w"
#eval test parseInstrPp "x := call f z   w" "x := call f z w"
#eval test parseInstrPp "o := cnstr 0   3   0" "o := cnstr 0 3 0"
#eval test parseInstrPp "set o 0 x" "set o 0 x"
#eval test parseInstrPp "x := get o 0" "x := get o 0"
#eval test parseInstrPp "sset o 8 x" "sset o 8 x"
#eval test parseInstrPp "x : Bool := sget o 24" "x : Bool := sget o 24"
#eval test parseInstrPp "x := closure f a" "x := closure f a"
#eval test parseInstrPp "x := closure f a b" "x := closure f a b"
#eval test parseInstrPp "x := apply f a" "x := apply f a"
#eval test parseInstrPp "x := Array sz c" "x := Array sz c"
#eval test parseInstrPp "arrayWrite a i v" "arrayWrite a i v"
#eval test parseInstrPp "x : object := arrayRead a i" "x : object := arrayRead a i"
#eval test parseInstrPp "x := sarray UInt32 sz c" "x := sarray UInt32 sz c"
#eval test parseInstrPp "arrayWrite a i v" "arrayWrite a i v"
#eval test parseInstrPp "x : UInt64 := arrayRead a i" "x : UInt64 := arrayRead a i"
#eval test parseInstrPp "inc x" "inc x"
#eval test parseInstrPp "dec x" "dec x"
#eval test parseInstrPp "decSref x" "decSref x"
#eval test parseInstrPp "free x" "free x"
#eval test parseInstrPp "x := call  f" "x := call f"
#eval test parseInstrPp "x:UInt32:=   arrayRead y   z" "x : UInt32 := arrayRead y z"

inductive Expr
| Add : Expr → Expr → Expr
| Num : Nat → Expr
| Var : String → Expr

open Expr

def eqExpr : Π a b : Expr, Decidable (a = b)
| (Var x)     (Var y)     := if h : x = y then isTrue (h ▸ rfl)
                             else isFalse (λ h', Expr.noConfusion h' (λ h', absurd h' h))
| (Var x)     (Num n)     := isFalse (λ h, Expr.noConfusion h)
| (Var x)     (Add e₁ e₂) := isFalse (λ h, Expr.noConfusion h)
| (Num n)     (Num m)     := if h : n = m then isTrue (h ▸ rfl)
                             else isFalse (λ h', Expr.noConfusion h' (λ h', absurd h' h))
| (Num n)     (Var y)     := isFalse (λ h, Expr.noConfusion h)
| (Num n)     (Add e₁ e₂) := isFalse (λ h, Expr.noConfusion h)
| (Add e₁ e₂) (Num n)     := isFalse (λ h, Expr.noConfusion h)
| (Add e₁ e₂) (Var y)     := isFalse (λ h, Expr.noConfusion h)
| (Add e₁ e₂) (Add e₃ e₄) :=
  match eqExpr e₁ e₃ with
  | isTrue h   := (match eqExpr e₂ e₄ with
                    | isTrue  h' := isTrue (h ▸ h' ▸ rfl)
                    | isFalse h' := isFalse (λ he, Expr.noConfusion he (λ h₁ h₂, absurd h₂ h')))
  | isFalse h := isFalse (λ he, Expr.noConfusion he (λ h₁ h₂, absurd h₁ h))

instance : DecidableEq Expr :=
{decEq := eqExpr}

def parseAtom (p : Parsec' Expr) : Parsec' Expr :=
(Var <$> lexeme var <?> "variable")
<|>
(Num <$> lexeme num <?> "numeral")
<|>
(paren p)

def parseAdd (p : Parsec' Expr) : Parsec' Expr :=
do l ← parseAtom p,
   (do symbol '+', r ← p, pure $ Add l r) <|> pure l

def parseExpr : Parsec' Expr :=
whitespace *> fix (λ F, parseAdd F) <* eoi

#eval test parseExpr "10" (Num 10)
#eval test parseExpr "(20)" (Num 20)
#eval test parseExpr "a" (Var "a")
#eval test parseExpr "(20 + a)" (Add (Num 20) (Var "a"))
#eval test parseExpr "  (20 + (a) + 2 ) " (Add (Num 20) (Add (Var "a") (Num 2)))

/- Failures -/
#print "Failure 1"
#eval test parseExpr "(20 +" (Num 0)
#print "---------"
#print "Failure 2"
#eval test parseExpr ""      (Num 0)
#print "---------"

namespace paperEx
#print "Failure 3"
def digit  : Parsec' Char := MonadParsec.digit <?> "digit"
def letter : Parsec' Char := MonadParsec.alpha <?> "letter"
def tst    : Parsec' Char := (digit <|> pure '0') *> letter
#eval test tst "*" 'a'
#print "---------"
end paperEx
