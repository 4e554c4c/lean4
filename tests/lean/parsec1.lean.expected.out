parsec1.lean:20:0: warning: using 'exit' to interrupt Lean
